### PR TITLE
Use coffeescript as a peerDependency

### PR DESCRIPTION
**Pull request by GeoffreyBooth**
Rather than `coffeeify` choosing a particular version of CoffeeScript and needing to be updated every time the upstream `coffeescript` module is updated, use `package.json`‘s [`peerDependencies`](https://nodejs.org/en/blog/npm/peer-dependencies/) key to require `coffeescript` to be installed but let the user specify the version. This adds support for both CoffeeScript 1.x and the upcoming 2.0.0.

@greghuc, I see you created a fork with something similar; do you mind testing this PR and letting me know if it works for you?

### CODE REVIEW

**Comment by greghuc** [comment]
@GeoffreyBooth I tried the PR in place of my fork, and all works fine. My package.json:

```
{
  "devDependencies": {
    "coffeeify": "github:GeoffreyBooth/coffeeify#peer",
    "coffeescript": "2.0.0-beta4",
    ...
  }
}

```

**Comment by jnordberg** [comment]
I don't like this, if coffee-script changes the compiler api everything breaks

**Comment by GeoffreyBooth** [comment]
> I don’t like this, if coffee-script changes the compiler api everything breaks

Yes, that’s the hazard of peer dependencies: https://nodejs.org/en/blog/npm/peer-dependencies/

And for that reason, you could specify a peer dependency version range, like `">= 1.9.1 < 3"`, instead of the `*` I put in this PR. But then you’re back to the reason I’m creating this PR in the first place: if CoffeeScript updates outside of the range you specify, you’ll need to update Coffeeify. Granted, that range buys you a lot of time, much more than locking to a particular version number; but you’re still just kicking the can down the road.

There have been [ten releases in just the last six months](http://coffeescript.org/v2/#changelog), with more to come. It’s a hassle to submit PRs to Coffeeify (and similar downstream tools) every time CoffeeScript updates; using the `package.json` `peerDependency` key solves that problem.

I’m a core contributor to CoffeeScript, and the developer organizing the CoffeeScript 2.0.0 release. We’re definitely not changing the Node API in 2.x, and probably not ever, for this very reason: we don’t want to break downstream tools like this one. If necessary, we may add _new_ options to the `options` argument passed to `compile`, but we certainly won’t change or break the existing ones. The API is documented [here](http://coffeescript.org/v2/#nodejs-usage) and hasn’t changed in years; I think it’s so unlikely to ever change, or at least to change in a backward-incompatible way, that I think a `peerDependency` range of `*` is safe.

**Comment by jnordberg** [comment]
Why then don't you separate the coffee-script package into a compiler api which in turn has the language itself as the peer dependency? That would allow you to update daily without forcing changes to packages downstream ;)

That said, if we make this a semver major along with the coffeescript 2.0 release I think it won't be too disruptive. And we need need a warning in the readme saying that after 2.0 you need to have your coffee-script version in the package.json as well.

**Comment by GeoffreyBooth** [comment]
> Why then don’t you separate the coffee-script package into a compiler api which in turn has the language itself as the peer dependency?

That’s actually not a bad idea; splitting out the CLI into its own package would probably be a good idea too. But at this point I don’t think anyone’s up for changes quite that radical, at least not in v2.

You wouldn’t need a warning in the readme (though one couldn’t hurt) because NPM itself will throw a warning on installation:
```
npm WARN coffeeify@2.2.0 requires a peer of coffeescript@* but none was installed.
```
And of course Coffeeify wouldn’t work until `coffeescript@*` were installed, so it’s not like the user wouldn’t notice.

**Comment by GeoffreyBooth** [comment]
P.S.: This is how `tsify` (TypeScript Browserify transform) does it:
https://github.com/TypeStrong/tsify/blob/master/package.json

```js
  "peerDependencies": {
    "browserify": ">= 10.x",
    "typescript": ">= 2.x"
  }
```

**Comment by d0liver** [comment]
Another possibility is to do what gulp-coffee does and allow the user to override the version of coffee being used by passing it in as an option. In that case the onus is on the user to provide a compliant API and it's their problem if it breaks. It also allows you to use any other compatible API this way so that coffeeify can support nonstandard compilers.

**Comment by GeoffreyBooth** [comment]
CoffeeScript 2.0.0 has been released.

**Comment by jnordberg** [comment]
Congrats! Why has the package name changed from coffee-script to coffeescript btw?

File: package.json [hunk 1]
```diff
@@ -1,15 +1,18 @@
 {
   "name": "coffeeify",
-  "version": "2.1.0",
-  "description": "browserify plugin for coffee-script with support for mixed .js and .coffee files",
+  "version": "2.2.0",
```
Comment (jnordberg) [inline package.json]: This is a semver major change, set it to 3.0.0

File: package.json [hunk 2]
```diff
@@ -1,15 +1,18 @@
...
+  "description": "browserify plugin for coffeescript with support for mixed .js and .coffee files",
   "main": "index.js",
   "dependencies": {
-    "coffee-script": "^1.12.0",
     "convert-source-map": "^1.3.0",
     "through2": "^2.0.0"
   },
+  "peerDependencies": {
+    "coffeescript": "*"
```
Comment (jnordberg) [inline package.json]: Set the peer dependency version to `<=2` so we can handle it if you ever feel like breaking the compiler api :)

**Comment by GeoffreyBooth** [comment]
Done. We moved to `coffeescript` because no one spells CoffeeScript with a hyphen.

**Comment by GeoffreyBooth** [comment]
Thanks!

**Pull request merged**